### PR TITLE
Feature/cache

### LINKED ISSUE #9: Server-side caching
We may want to look into adding a server-side cache of some kind. We probably want to make this work with dask and dask-distributed for keys explicitly computed, e.g.:

https://github.com/jhamman/xpublish/blob/36befed89d880b8d249b1c587dd994e40a3e76a0/xpublish/rest.py#L299-L300

I know dask has some experimental caching functionality (https://docs.dask.org/en/latest/caching.html) as well as an old companion project [cachey](https://github.com/dask/cachey). It is possible we don't really need anything since dask's caching could be turned on by the data publisher:

```python

import xarray as xr
import xpublish
from dask.cache import Cache

cache = Cache(2e9)  # Leverage two gigabytes of memory
cache.register()    # Turn cache on globally

ds = xr.open_mfdataset(...)
...
ds.rest.serve()

```

I'm also curious how dask's caching functionality works with distributed.

### PR DESCRIPTION
**Pull request by jhamman**
closes #9 (maybe)

This PR is a proof of concept for server-side caching. It uses cachey to cache key responses. 

cc @andersy005

### CODE REVIEW

File: xpublish/rest.py [hunk 1]
```diff
@@ -97,18 +100,26 @@ def zmetadata_json(self):
...
-            data_chunk.tobytes(), filters=arr_meta['filters'], compressor=arr_meta['compressor'],
-        )
-        return Response(echunk, media_type='application/octet-stream')
+                echunk = _encode_chunk(
+                    data_chunk.tobytes(),
+                    filters=arr_meta['filters'],
+                    compressor=arr_meta['compressor'],
+                )
+                response = Response(echunk, media_type='application/octet-stream')
+            self._cache.put(cache_key, response, max(1, int(ct.time)), len(echunk))
```
Comment (andersy005) [inline xpublish/rest.py]: This is neat! So, is this being cached in the same memory pool used by the server?

Comment (jhamman) [reply xpublish/rest.py]: Yes, exactly! Which I think is what we need anyways.

File: xpublish/rest.py [hunk 2]
```diff
@@ -222,11 +250,11 @@ def info():
             return self._info()
 
         @self._app.get('/dict')
-        def to_dict(data: bool = False):
-            return self._obj.to_dict(data=data)
+        def to_dict():
+            return self._obj.to_dict(data=False)
```
Comment (jhamman) [inline xpublish/rest.py]: This is a breaking change but I think it is needed for safety's sake. Exposing the full dataset as a single string could lead to bring down the server with a memory error.

File: xpublish/rest.py [hunk 3]
```diff
@@ -49,6 +56,44 @@ def __init__(self, xarray_obj):
...
+        ----------
+        cache_kws : dict
+            Dictionary of keyword arguments to be passed to ``cachey.Cache()``
+        app_kws : dict
+            Dictionary of keyword arguments to be passed to
+            ``fastapi.FastAPI()``
+
+        Notes
+        -----
+        This method can only be inoked once.
```
Comment (jhamman) [inline xpublish/rest.py]: typo

File: xpublish/rest.py [hunk 4]
```diff
@@ -49,6 +56,44 @@ def __init__(self, xarray_obj):
...
+        if self._initialized:
+            raise RuntimeError('This accessor has already been initialized')
+        self._initialized = True
+
+        # update kwargs
+        if cache_kws is not None:
+            self._cache_kws.update(cache_kws)
+        if app_kws is not None:
+            self._app_kws.update(app_kws)
+        return self
```
Comment (jhamman) [inline xpublish/rest.py]: @lsetiawan - please give me your thoughts on this change. I've made `init_app()` a private method and instead exposed the accessor's `__call__` method as the preferred way to set options. So now, you're app configuration would go like this:

```python
ds.rest(app_kws=dict(title='foo')).serve()
```

Comment (lsetiawan) [reply xpublish/rest.py]: This looks really great! i like the cleanness of this. Nice work!

Comment (andersy005) [reply xpublish/rest.py]: > ds.rest(app_kws=dict(title='foo')).serve()

This is great! I didn't know that this is the proper way to set options when using the accessor until now :)

Comment (jhamman) [reply xpublish/rest.py]: I don't think we have a "proper" way defined but this is fairly sane.

Comment (andersy005) [reply xpublish/rest.py]: This would be great to have in the xarray documentation.

Comment (jhamman) [reply xpublish/rest.py]: +1:  https://github.com/pydata/xarray/issues/3829

**Pull request merged**